gnc/sub8_perception/nodes/buoy.cc

```diff
@@ -319,8 +341,99 @@ sub::statistical_outlier_filter<sub::PointXYZT>(scene_buffer, current_cloud,
+
+}
+
+cv::Mat torpedo_board_segmentation(const cv::Mat &image);
+
+
+void Sub8TorpedoBoardDetector::image_callback(const sensor_msgs::ImageConstPtr &image_msg,
+                               const sensor_msgs::CameraInfoConstPtr &info_msg){
+  cv_bridge::CvImagePtr input_bridge;
+  cv::Mat current_image, segmented_board_left, segmented_board_right;
+  try {
+    input_bridge = cv_bridge::toCvCopy(image_msg, sensor_msgs::image_encodings::BGR8);
+    current_image = input_bridge->image;
+  } catch (cv_bridge::Exception &ex) {
+    ROS_ERROR("[torpedo_board] cv_bridge: Failed to convert image");
+    return;
+  }
+  cam_model.fromCameraInfo(info_msg);
+  segmented_board_left = board_segmentation(current_image);
+  cv::waitKey(1);
+}
+
+
+void Sub8TorpedoBoardDetector::determine_torpedo_board_position(const image_geometry::PinholeCameraModel &cam_model,
+                                             const cv::Mat &image_raw){
+
+}
+
+
+bool Sub8TorpedoBoardDetector::request_torpedo_board_position(sub8_msgs::VisionRequest::Request &req,
+                                           sub8_msgs::VisionRequest::Response &resp){
+  return true;
+}
+
+
+cv::Mat Sub8TorpedoBoardDetector::board_segmentation(const cv::Mat &image){
+
+  // Preprocessing
+  cv::Mat processing_size_image, hsv_image, hue_blurred, sat_blurred;
+  std::vector<cv::Mat> hsv_channels;
+  cv::resize(image, processing_size_image, cv::Size(0,0), 0.5, 0.5);
+  cv::cvtColor(processing_size_image, hsv_image, CV_BGR2HSV);
+  cv::split(hsv_image, hsv_channels);
+  cv::medianBlur(hsv_channels[0], hue_blurred, 5); // Magic num: 5 (might need tuning)
+  cv::medianBlur(hsv_channels[1], sat_blurred, 5);
+  // cv::imshow("hue median filtered", hue_blurred);
+
+  // Histogram parameters
+  int hist_size = 256;    // bin size
+  float range[] = { 0, 255 };
+  const float *ranges[] = { range };
+
+  // Segment out torpedo board
+  cv::Mat threshed_hue, threshed_sat, segmented_board;
+  int target_yellow = 20;
+  int target_saturation = 180;
+  sub::statistical_image_segmentation(hue_blurred, threshed_hue, hist_size, ranges,
+                                      target_yellow, "Hue", 6, 0.5, 0.5);
+  sub::statistical_image_segmentation(sat_blurred, threshed_sat, hist_size, ranges,
+                                      target_saturation, "Saturation", 6, 0.1, 0.1);
+  segmented_board = threshed_hue / 2.0 + threshed_sat / 2.0;
+  cv::threshold(segmented_board, segmented_board, 255*0.75, 255, cv::THRESH_BINARY);
+  
+#ifdef VISUALIZE
+  cv::imshow("segmented board", segmented_board);
+#endif
+
+  cv::resize(segmented_board, segmented_board, cv::Size(0,0), 2, 2);
+  return segmented_board;
+}
+
+
 int main(int argc, char **argv) {
   ros::init(argc, argv, "pcl_slam");
-  boost::shared_ptr<Sub8BuoyDetector> sub8_buoys(new Sub8BuoyDetector());
+  ROS_INFO("Initializing node /pcl_slam");
+  // boost::shared_ptr<Sub8BuoyDetector> sub8_buoys(new Sub8BuoyDetector());
```

Comment:
cv_tools is intended to be the header for more than one cv_*.cc, much like pcl_tools. Originally, cv_utils had just a handful of miscellaneous utility functions